src/storage-operators/src/s3_oneshot_sink.rs

```diff
@@ -45,21 +47,33 @@ pub fn copy_to<G, F>(
     connection_details: S3UploadInfo,
     connection_context: ConnectionContext,
     aws_connection: AwsConnection,
+    sink_id: GlobalId,
     connection_id: GlobalId,
     onetime_callback: F,
 ) where
     G: Scope<Timestamp = Timestamp>,
     F: FnOnce(Result<u64, String>) -> () + 'static,
 {
-    let scope = input_collection.scope();
+    let mut scope = input_collection.scope();
     let worker_id = scope.index();
+    let is_leader = usize::cast_from((sink_id, "leader").hashed()) % scope.peers() == worker_id;
 
-    let mut builder = AsyncOperatorBuilder::new("CopyToS3".to_string(), scope);
+    let mut builder = AsyncOperatorBuilder::new("CopyToS3".to_string(), scope.clone());
 
     let mut input_handle = builder.new_disconnected_input(&input_collection.inner, Pipeline);
     let mut error_handle = builder.new_disconnected_input(&err_collection.inner, Pipeline);
 
-    builder.build(move |_caps| async move {
+    // Signal mechanism from leader to all workers to start the upload after it has verified
+    // the integrity of the S3 bucket path. We check the S3 bucket path from a single worker
+    // to avoid a race between checking the path vs workers uploading objects to the path.
+    let (mut start_handle, start_stream) = builder.new_output();
+    let (start_feedback_handle, start_feedback_stream) = scope.feedback(Default::default());
+    let mut start_input = builder.new_disconnected_input(&start_feedback_stream, Pipeline);
```

Comment:
```suggestion
    // Signal mechanism from leader to all workers to start the upload after it has verified
    // the integrity of the S3 bucket path. We check the S3 bucket path from a single worker
    // to avoid a race between checking the path vs workers uploading objects to the path.
    //
    // A `()` datum along this edge indicates success. A `[]` frontier indicates failure. Note that this
    // is disconnected and has a `0` summary, as its unrelated to the frontier of the output.
    let (mut start_handle, start_stream) = builder.new_output();
    let (start_feedback_handle, start_feedback_stream) = scope.feedback(Default::default());
    let mut start_input = builder.new_disconnected_input(&start_feedback_stream, Pipeline);
```